tests/unit/adapters/osf-adapter-test.js

```diff
@@ -1,12 +1,401 @@
+        'transform:links', 'transform:embed'
+    ],
+    beforeEach() {
+        manualSetup(this.container);
+    }
+});
+
+test('#buildURL appends a trailing slash if missing', function(assert) {
+    var url = 'https://api.osf.io/v2/users/me';
+    this.stub(
+        DS.JSONAPIAdapter.prototype,
+        'buildURL',
+        function() {
+            return url;
+        }
+    );
+    let adapter = this.subject();
+    let user = FactoryGuy.make('user');
+    let result = adapter.buildURL(
+        'user',
+        'me',
+        user._internalModel.createSnapshot(),
+        'findRecord'
+    );
+    assert.notEqual(url, result);
+    assert.equal(result.slice(-1), '/');
+});
+
+test('#buildURL uses relationship links if available for delete, update, and find', function(assert) {
+    let url = 'https://api.osf.io/v2/users/me/rel/';
+    let adapter = this.subject();
+    let user = FactoryGuy.make('user', {
+        links: {
+            self: url
+        }
+    });
+    ['delete', 'update', 'find'].forEach(verb => {
+        let result = adapter.buildURL(
+            'user',
+            'me',
+            user._internalModel.createSnapshot(),
+            `${verb}Record`
+        );
+        assert.equal(url, result);
+    });
+});
+
+test('#buildURL uses snapshot.adapterOptions.url if available', function(assert) {
+    let url = 'https://api.osf.io/v2/users/me/rel/';
+    let adapter = this.subject();
+    let user = FactoryGuy.make('user', {
+        links: null
+    });
+
+    let result = adapter.buildURL(
+        'user',
+        'me',
+        user._internalModel.createSnapshot({
+            adapterOptions: {
+                url: url
+            }
+        }),
+        'createRecord'
+    );
+    assert.equal(url, result);
+});
+
+test('#_buildRelationshipURL uses relationshipLinks', function(assert) {
+    let url = 'https://api.osf.io/v2/users/me/nodes/';
+    let adapter = this.subject();
+    let user = FactoryGuy.make('user', {
+        links: {
+            relationships: {
+                nodes: {
+                    links: {
+                        related: {
+                            href: url
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    let result = adapter._buildRelationshipURL(
+        user._internalModel.createSnapshot({
+            adapterOptions: {
+                url: url
+            }
+        }),
+        'nodes'
+    );
+    assert.equal(url, result);
+});
+
+test('#_createRelated maps over each createdSnapshots and adds records to the parent\'s canonical state', function(assert) {
+    this.inject.service('store');
+    let store = this.store;
+
+    let node = FactoryGuy.make('node');
+    Ember.run.begin();
+    let contributors = [
+        store.createRecord('contributor', {
+            title: 'Foo'
+        }),
+        store.createRecord('contributor', {
+            title: 'Bar'
+        })
+    ];
+    Ember.run.end();
```

Comment:
What is the advantage of manually specifying begin and end? Alternately, you can run an entire block as `Ember.run(function() {})`